routing/index_graph_serializer.hpp

```diff
@@ -0,0 +1,411 @@
+    Joint::Id Read(BitReader<Source> & reader)
+    {
+      uint8_t const bit = reader.Read(1);
+      if (bit == kRepeatJointIdBit)
+      {
+        uint32_t const delta = ReadDelta(reader);
+        if (delta > m_count)
+          MYTHROW(RoutingException, ("Joint id delta", delta, "> count =", m_count));
+
+        return m_count - delta;
+      }
+
+      return m_count++;
+    }
+
+    Joint::Id GetCount() const { return m_count; }
+
+  private:
+    Joint::Id m_count;
+  };
+
+  // PointsFilter has two goals:
+  //
+  // 1. Deserialization skips some sections.
+  //    Therefore Joint::Id is not a continuous series of number [0, N).
+  //    PointsFilter converts joint ids to continuous series of number [0, N).
+  //
+  // 2. Some joints connects roads from different models.
+  //    E.g. 2 roads joint: all vehicles road + pedestrian road.
+  //    If we loads in car model, pedestrian road is skipped,
+  //    so joint would have only point from all vehicles road.
+  //    PointsFilter filters such redundant joints.
+
+  class PointsFilter final
+  {
+  public:
+    PointsFilter(IndexGraph & graph, Joint::Id numJoints) : m_graph(graph)
+    {
+      m_entries.assign(numJoints, {kEmptyEntry, 0});
+    }
+
+    void Push(Joint::Id jointIdInFile, RoadPoint const & rp);
+    Joint::Id GetCount() const { return m_count; }
+
+  private:
+    static uint32_t constexpr kEmptyEntry = numeric_limits<uint32_t>::max();
+    static uint32_t constexpr kPushedEntry = kEmptyEntry - 1;
+
+    // Joint::Id and point id are stored in same space.
+    // Therefore size of Joint::Id and point id should be equal.
+    static_assert(sizeof(Joint::Id) == sizeof(uint32_t),
+                  "Can't store Joint::Id in m_entries[i].second");
+
+    IndexGraph & m_graph;
+    Joint::Id m_count = 0;
+    vector<pair<uint32_t, uint32_t>> m_entries;
+  };
+
+  class SectionSerializer final
+  {
+  public:
+    explicit SectionSerializer(VehicleMask mask) : m_mask(mask) {}
+
+    size_t GetBufferSize() const { return m_buffer.size(); }
+    VehicleMask GetMask() const { return m_mask; }
+    size_t GetNumRoads() const { return m_featureIds.size(); }
+
+    void AddRoad(uint32_t featureId) { m_featureIds.push_back(featureId); }
+    void SortRoads() { sort(m_featureIds.begin(), m_featureIds.end()); }
+    void PreSerialize(IndexGraph const & graph, unordered_map<uint32_t, VehicleMask> const & masks,
+                      JointIdEncoder & jointEncoder);
+
+    template <class Sink>
+    void Flush(Sink & sink)
+    {
+      sink.Write(m_buffer.data(), m_buffer.size());
+      m_buffer.clear();
+    }
+
+  private:
+    VehicleMask const m_mask;
+    vector<uint32_t> m_featureIds;
+    vector<uint8_t> m_buffer;
+  };
+
+  template <typename Sink>
+  static void WriteGamma(BitWriter<Sink> & writer, uint32_t value)
+  {
+    ASSERT_NOT_EQUAL(value, 0, ());
+
+    bool const success = coding::GammaCoder::Encode(writer, static_cast<uint64_t>(value));
+    if (!success)
```

Comment:
GammaCoder ALWAYS return true when the argument is not zero. It's a GammaCoder contract.